fix(api): Prioritize mc alarms over errors

sfoster1 opened the pull request:
The error message returned by the motor controller after a hard halt is
"error:Alarm lock". This is meant to be sensed by logic that raises a
different kind of exception (SmoothieAlarm) from the kind raised by for
instance an unexpected limit switch actuation or malformed
response (SmoothieError). Unfortunately, the _way_ this is sensed is
that SmoothieAlarm looks for the keyword "alarm" and SmoothieError looks
for the keyword "error", and that message contains both, and the error
case is checked first.

The fix is to swap the order of the two. The fix could also be to
special case the message checking, but let's see if keeping it simple
works first.

## Testing
This is unit tested and will do what it says on the tin, but it's not clear to me if this is a safe thing to do in all circumstances on hardware. We should test this by running it on a robot and

- Cancel a running protocol (or do something else that causes a hard halt) and make sure the error recovery in case of a SmoothieAlarm actually, like, works, since it's probably never occurred before
- Run a protocol and tap a limit switch to cause a `HardLimitError` and make sure it's still a `SmoothieError`
- Hold down the switch during a jog or non-protocol-execution motion and check the same

sfoster1 committed:
fix(api): Prioritize mc alarms over errors

The error message returned by the motor controller after a hard halt is
"error:Alarm lock". This is meant to be sensed by logic that raises a
different kind of exception (SmoothieAlarm) from the kind raised by for
instance an unexpected limit switch actuation or malformed
response (SmoothieError). Unfortunately, the _way_ this is sensed is
that SmoothieAlarm looks for the keyword "alarm" and SmoothieError looks
for the keyword "error", and that message contains both, and the error
case is checked first.

The fix is to swap the order of the two. The fix could also be to
special case the message checking, but let's see if keeping it simple
works first.

mcous reviewed (approved):
LGTM based on code reading. Away from robot at the moment, but can test later today if needed

SyntaxColoring commented:
> Run a protocol and tap a limit switch to cause a HardLimitError and make sure it's still a SmoothieError

I'm not sure if I'm testing this right, but this might not be the case?

I ran a protocol and reached in and held down the +X limit switch. The robot stopped and then homed. I see these errors in the `GET /run` response:

```json
        {
            "id": "a3297d01-2b14-4a7b-a234-11b42aec3c12",
            "createdAt": "2022-01-11T15:41:07.272963+00:00",
            "status": "failed",
            "current": true,
            "actions": [
                {
                    "id": "7f97d037-cd50-466d-81b5-faa208c94c63",
                    "createdAt": "2022-01-11T15:41:14.393300+00:00",
                    "actionType": "play"
                }
            ],
            "commands": [
                [...]
                {
                    "id": "command.DISPENSE-2",
                    "commandType": "custom",
                    "status": "succeeded"
                },
                {
                    "id": "command.RETURN_TIP-2",
                    "commandType": "custom",
                    "status": "failed",
                    "errorId": "13d98c6e-2ff2-44b8-8d45-a68926b12513"
                },
                {
                    "id": "command.DROP_TIP-2",
                    "commandType": "dropTip",
                    "status": "failed",
                    "errorId": "20010b69-781a-46f5-99b7-70e19a00aa32"
                }
            ],
            "errors": [
                {
                    "id": "20010b69-781a-46f5-99b7-70e19a00aa32",
                    "errorType": "LegacyContextCommandError",
                    "createdAt": "2022-01-11T15:42:21.809594+00:00",
                    "detail": "SmoothieAlarm: None returned ALARM: Hard limit -X"
                },
                {
                    "id": "13d98c6e-2ff2-44b8-8d45-a68926b12513",
                    "errorType": "LegacyContextCommandError",
                    "createdAt": "2022-01-11T15:42:21.812395+00:00",
                    "detail": "SmoothieAlarm: None returned ALARM: Hard limit -X"
                },
                {
                    "id": "564028e9-88a7-48e2-8e27-89447e1a97c6",
                    "errorType": "SmoothieAlarm",
                    "createdAt": "2022-01-11T15:42:21.819024+00:00",
                    "detail": "SmoothieAlarm: None returned ALARM: Hard limit -X"
                }
            ],
            "pipettes": [...]
            "labware": [...]
            "labwareOffsets": [],
            "protocolId": "8dfbc3b3-92ff-4a41-83b1-f9317cc0283f"
        }
```

Laura-Danielle reviewed (approved):
Looks good as long as it's tested on a robot.

sfoster1 committed:
Clear hard halting in reset from error

Should prevent a latched hard-halting condition from occurring that
would incorrectly not transform limit switch alarms to errors.

SyntaxColoring commented:
With the latest commit, `GET /runs` now shows `SmoothieError` as specified in the review requests.

However, I noticed this behavior while jogging in deck calibration:

1. Start deck calibration and click through the dialogs until you have the opportunity to jog.
2. Hold the +X limit switch down.
4. While holding down, the limit switch, jog to the right. Release the limit switch.
5. **The robot will home in the Z, A, and X, but *not* Y.** This is weird to me?
6. **Try jogging to the left. The robot will instead home all axes, including pipette plungers.** This is also weird to me?
7. Jog to the left again. This time, the robot doesn't home, and the jog proceeds as normal. **No errors are propagated to the Opentrons App.**

It looks like this behavior is also present on `edge` b39d2f9a427baba887e751eb198577ed0ca93205.

sfoster1 committed:
add tests for halt behavior

Laura-Danielle reviewed (approved):
The tests themselves look good. I think it would be helpful to have some more context about the expected physical robot behavior change in the test doc strings.

If I understand correctly from the thread/this PR, we effectively don't want to trigger a home when we hit a limit switch?

sfoster1 commented:
> The tests themselves look good. I think it would be helpful to have some more context about the expected physical robot behavior change in the test doc strings.
> 
> If I understand correctly from the thread/this PR, we effectively don't want to trigger a home when we hit a limit switch?

We do want to - we want to keep behavior unchanged. The behavior @SyntaxColoring noted above is as intended, even if what's intended is a little weird, and is present in edge. The strange homing behavior is because of required homing sequences and trying to home just the erroring axis to avoid e.g. overaspiration - just homing X requires ZA, then Y's homing flag is reset so it homes Y but that requires homing X and thus ZA.
